File: backend/app/modules/synoptic/calculations.py
"""
Módulo de Cálculos Meteorológicos
Basado en el Manual de Claves OMM N°306 (FM-12 SYNOP)
"""

import math
from typing import Dict, Optional, Any

# =============================================================================
# DATOS DE ESTACIONES (IIiii → Corrección de Altura)
# =============================================================================

STATIONS = {
    "78451": {"name": "Monte Cristi", "ch": 0.5, "lat": 19.8499, "lon": -71.6535, "h": 8},
    "MDCY": {"name": "Catey", "ch": 0.8, "lat": 19.267, "lon": -69.7337, "h": 4},
    "78457": {"name": "Puerto Plata", "ch": 1.0, "lat": 19.7542, "lon": -70.5632, "h": 16},
    "78482": {"name": "Barahona", "ch": 1.2, "lat": 18.24861, "lon": -71.12288, "h": 19.5},
    "78467": {"name": "Sabana de la mar", "ch": 1.2, "lat": 19.0527, "lon": -69.3888, "h": 11},
    "78464": {"name": "Cabrera", "ch": 1.5, "lat": 19.6444, "lon": -69.9063, "h": 18},
    "78486": {"name": "Central", "ch": 1.6, "lat": 18.4734, "lon": -69.8705, "h": 14},
    "78485": {"name": "Las américas", "ch": 2.0, "lat": 18.4331, "lon": -69.6796, "h": 7},
    "78479": {"name": "Punta Cana", "ch": 2.0, "lat": 18.546, "lon": -68.3594, "h": 7},
    "78484": {"name": "El Higüero", "ch": 3.5, "lat": 18.57696, "lon": -69.98158, "h": 27},
    "78466": {"name": "Arroyo Barril", "ch": 3.5, "lat": 19.2005, "lon": -69.43144, "h": 49.4},
    "78480": {"name": "Jimaní", "ch": 4.8, "lat": 18.4928, "lon": -71.853, "h": 45},
    "78473": {"name": "Bayaguana", "ch": 6.0, "lat": 18.7422, "lon": -69.6308, "h": 53},
    "78488": {"name": "La Romana", "ch": 8.5, "lat": 18.4485, "lon": -68.9093, "h": 62},
    "78460": {"name": "Santiago", "ch": 19.8, "lat": 19.4031, "lon": -70.5978, "h": 170},
}


def get_station_info(station_id: str) -> Optional[Dict]:
    """Obtiene información de una estación por su ID (IIiii)."""
    # Buscar por ID exacto o por los últimos 5 dígitos
    station_id = station_id.strip().upper()
    if station_id in STATIONS:
        return STATIONS[station_id]
    # Buscar solo por número (sin prefijo)
    for key, value in STATIONS.items():
        if key.endswith(station_id) or station_id.endswith(key):
            return value
    return None


# =============================================================================
# CÁLCULOS METEOROLÓGICOS
# =============================================================================

def calcular_h26(temp_humedo: float) -> Optional[float]:
    """
    Calcula H26 - Presión de vapor de saturación a temperatura húmeda.
    Fórmula Magnus-Tetens.
    """
    if temp_humedo + 243.5 == 0:
        return None
    return 6.112 * math.exp((17.67 * temp_humedo) / (temp_humedo + 243.5))


def calcular_h27(temp_humedo: float) -> float:
    """Calcula H27 - Constante psicrométrica."""
    return 0.00066 * 1000 * (1 + 0.00115 * temp_humedo)


def calcular_h28_asumido(temp_seco: float) -> Optional[float]:
    """Calcula H28 - Presión de vapor de saturación a temperatura seca."""
    if temp_seco + 243.5 == 0:
        return None
    return 6.112 * math.exp((17.67 * temp_seco) / (temp_seco + 243.5))


def calcular_tension_vapor(temp_seco: float, temp_humedo: float) -> Optional[float]:
    """
    Calcula la tensión de vapor usando la fórmula psicrométrica.
    TV = H26 - H27 * (Ts - Th)
    """
    h26 = calcular_h26(temp_humedo)
    h27 = calcular_h27(temp_humedo)
    if h26 is None:
        return None
    return h26 - h27 * (temp_seco - temp_humedo)


def calcular_humedad_relativa(tension_vapor: float, temp_seco: float) -> Optional[float]:
    """
    Calcula la humedad relativa.
    HR = (TV / H28_asumido) * 100
    """
    h28_asumido = calcular_h28_asumido(temp_seco)
    if h28_asumido is None or h28_asumido == 0:
        return None
    return (tension_vapor / h28_asumido) * 100


def calcular_punto_rocio(temp_seco: float, humedad_relativa_porc: float) -> Optional[float]:
    """Calcula el punto de rocío usando fórmula polinomial."""
    try:
        factor_humedad = 1 - (0.01 * humedad_relativa_porc)
        termino1 = (14.55 + 0.114 * temp_seco) * factor_humedad
        termino2 = ((2.5 + 0.007 * temp_seco) * factor_humedad) ** 3
        termino3 = (15.9 + 0.117 * temp_seco) * (factor_humedad ** 14)
        return temp_seco - termino1 - termino2 - termino3
    except Exception:
        return None


# =============================================================================
# CODIFICACIÓN SYNOP (WMO FM-12)
# =============================================================================

def format_temperature_group(temp: float) -> str:
    """
    Formatea grupo de temperatura 1snTTT.
    sn = 0 para temp >= 0, sn = 1 para temp < 0
    TTT = temperatura × 10 (sin signo)
    Ejemplo: 23.3°C → "10233", -23.3°C → "11233"
    """
    if temp is None or math.isnan(temp):
        return ""
    sign = "0" if temp >= 0 else "1"
    abs_temp = abs(temp)
    ttt = str(round(abs_temp * 10)).zfill(3)
    return f"1{sign}{ttt}"


def format_dew_point_group(dew_point: float) -> str:
    """
    Formatea grupo de punto de rocío 2snTdTdTd.
    sn = 0 para temp >= 0, sn = 1 para temp < 0
    Ejemplo: 21.5°C → "20215", -21.5°C → "21215"
    """
    if dew_point is None or math.isnan(dew_point):
        return ""
    sign = "0" if dew_point >= 0 else "1"
    abs_dew_point = abs(dew_point)
    ttt = str(round(abs_dew_point * 10)).zfill(3)
    return f"2{sign}{ttt}"


def format_pressure_group(pressure: float) -> str:
    """
    Formatea grupo de presión a nivel del mar 4PPPP.
    PPPP = presión × 10 mod 10000 (últimos 4 dígitos)
    Ejemplo: 1017.3 hPa → "40173", 998.5 hPa → "49985"
    """
    if pressure is None or math.isnan(pressure):
        return ""
    pppp = str(round(pressure * 10) % 10000).zfill(4)
    return f"4{pppp}"


def calcular_tendencia_a(diferencia_p3: float) -> str:
    """
    Calcula el código 'a' para el grupo 5aPPP según la tendencia de presión.
    
    Tendencia positiva (presión subiendo):
        0: +0.1 a +0.5 hPa
        1: +0.6 a +1.4 hPa
        2: +1.5 a +1.9 hPa
        3: +2.0 hPa o más
    
    Sin cambio:
        4: 0 hPa
    
    Tendencia negativa (presión bajando):
        5: -0.1 a -0.5 hPa
        6: -0.6 a -1.4 hPa
        7: -1.5 a -1.9 hPa
        8: -2.0 hPa o menos
    """
    if diferencia_p3 is None or math.isnan(diferencia_p3):
        return ""
    
    # Redondear a 1 decimal
    dif = round(diferencia_p3, 1)
    
    # Sin cambio
    if dif == 0:
        return "4"
    
    abs_dif = abs(dif)
    
    # Tendencia positiva (subiendo)
    if dif > 0:
        if 0.1 <= abs_dif <= 0.5:
            return "0"
        elif 0.6 <= abs_dif <= 1.4:
            return "1"
        elif 1.5 <= abs_dif <= 1.9:
            return "2"
        elif abs_dif >= 2.0:
            return "3"
    
    # Tendencia negativa (bajando)
    if dif < 0:
        if 0.1 <= abs_dif <= 0.5:
            return "5"
        elif 0.6 <= abs_dif <= 1.4:
            return "6"
        elif 1.5 <= abs_dif <= 1.9:
            return "7"
        elif abs_dif >= 2.0:
            return "8"
    
    return "4"  # Por defecto, sin cambio


def format_pressure_tendency_group(diferencia_p3: float) -> str:
    """
    Formatea el grupo 5aPPP (tendencia de presión en 3 horas).
    a = código de tendencia (0-8)
    PPP = magnitud del cambio × 10
    """
    if diferencia_p3 is None or math.isnan(diferencia_p3):
        return ""
    
    a = calcular_tendencia_a(diferencia_p3)
    ppp = str(round(abs(diferencia_p3) * 10)).zfill(3)
    return f"5{a}{ppp}"


def should_include_precipitation(ir: str) -> bool:
    """
    Determina si el grupo 6RRR debe incluirse según Ir.
    Ir = 0, 1, 2: incluir; Ir = 3, 4: omitir
    """
    try:
        ir_num = int(ir)
        return 0 <= ir_num <= 2
    except (ValueError, TypeError):
        return True  # Por defecto incluir


def should_include_weather(ix: str) -> bool:
    """
    Determina si el grupo 7ww debe incluirse según Ix.
    Ix = 1, 4: incluir; Ix = 2, 3, 5, 6, 7: omitir
    """
    try:
        ix_num = int(ix)
        return ix_num in (1, 4)
    except (ValueError, TypeError):
        return True  # Por defecto incluir


def format_pressure_24h_group(diferencia_p24: float) -> str:
    """
    Formatea el grupo 58/59 P24P24P24 (cambio de presión en 24 horas).
    58 = presión subió (diferencia positiva)
    59 = presión bajó (diferencia negativa)
    P24P24P24 = magnitud del cambio × 10
    
    Ejemplo: +3.5 hPa → "58035", -2.1 hPa → "59021"
    """
    if diferencia_p24 is None or math.isnan(diferencia_p24):
        return ""
    
    # Redondear a 1 decimal
    dif = round(diferencia_p24, 1)
    
    # Determinar prefijo según si subió o bajó
    if dif >= 0:
        prefix = "58"
    else:
        prefix = "59"
    
    # PPP = magnitud × 10
    ppp = str(round(abs(dif) * 10)).zfill(3)
    return f"{prefix}{ppp}"


def format_humidity_group(humedad_relativa: float) -> str:
    """
    Formatea el grupo 29UUU (humedad relativa).
    UUU = humedad relativa (3 dígitos)
    
    Ejemplo: 71% → "29071", 100% → "29100"
    """
    if humedad_relativa is None or math.isnan(humedad_relativa):
        return ""
    
    # Redondear y asegurar 3 dígitos
    hr = round(min(100, max(0, humedad_relativa)))
    uuu = str(hr).zfill(3)
    return f"29{uuu}"


# =============================================================================
# FUNCIÓN PRINCIPAL DE CÁLCULO
# =============================================================================

def realizar_calculos(data: Dict[str, Any]) -> Dict[str, Any]:
    """
    Realiza todos los cálculos meteorológicos y genera códigos SYNOP.
    
    Args:
        data: Diccionario con los datos del formulario
        
    Returns:
        Diccionario con todos los resultados calculados
    """
    results = {
        # Cálculos de temperatura
        "tension_vapor": "",
        "humedad_relativa": "",
        "punto_rocio": "",
        "diferencia": "",
        
        # Códigos SYNOP
        "grupo_1sn_ttt": "",
        "grupo_2sn_td": "",
        "grupo_4pppp": "",
        "grupo_5appp": "",
        "grupo_58_59_p24": "",  # 58/59 P24P24P24 (24h pressure change)
        "grupo_29uuu": "",      # 29UUU (humidity)
        
        # Cálculos de presión
        "p3_let": "",
        "p24_let": "",
        "p3_dif": "",
        "p24_dif": "",
        "pres_nmm": "",
        "correc_alt": "",
        
        # Indicadores condicionales
        "include_precipitation": True,
        "include_weather": True,
        
        # Información de estación
        "station_info": None,
        
        # Errores
        "error_message": ""
    }
    
    # =========================================================================
    # LOOKUP DE ESTACIÓN (IIiii → Correc. Alt.)
    # =========================================================================
    station_id = data.get("station_id", "").strip()
    if station_id:
        station_info = get_station_info(station_id)
        if station_info:
            results["station_info"] = station_info
            results["correc_alt"] = str(station_info["ch"])
    
    # Si no hay station_id pero hay correc_alt manual, usarlo
    if not results["correc_alt"]:
        results["correc_alt"] = data.get("correc_alt", "")
    
    # =========================================================================
    # CÁLCULOS DE TEMPERATURA
    # =========================================================================
    ts_str = data.get("ts", "")
    th_str = data.get("th", "")
    
    temp_seco = None
    punto_rocio = None
    
    if ts_str:
        try:
            temp_seco = float(ts_str)
            # Generar grupo 1snTTT
            results["grupo_1sn_ttt"] = format_temperature_group(temp_seco)
        except ValueError:
            pass
    
    if ts_str and th_str:
        try:
            temp_seco = float(ts_str)
            temp_humedo = float(th_str)
            
            # Validación: Th no puede ser mayor que Ts
            if temp_humedo > temp_seco:
                results["error_message"] = "Error: Th no puede ser mayor que Ts."
            else:
                # Calcular tensión de vapor
                tension_vapor = calcular_tension_vapor(temp_seco, temp_humedo)
                
                if tension_vapor is not None:
                    results["tension_vapor"] = f"{tension_vapor:.1f}"
                    
                    # Calcular humedad relativa
                    humedad_relativa = calcular_humedad_relativa(tension_vapor, temp_seco)
                    
                    if humedad_relativa is not None:
                        results["humedad_relativa"] = f"{min(100, humedad_relativa):.0f}"
                        
                        # Calcular punto de rocío
                        punto_rocio = calcular_punto_rocio(temp_seco, humedad_relativa)
                        
                        if punto_rocio is not None:
                            results["punto_rocio"] = f"{punto_rocio:.1f}"
                            # Generar grupo 2snTdTdTd
                            results["grupo_2sn_td"] = format_dew_point_group(punto_rocio)
                        
                        # Generar grupo 29UUU (humedad)
                        results["grupo_29uuu"] = format_humidity_group(humedad_relativa)
                
                # Calcular diferencia
                diferencia = temp_seco - temp_humedo
                results["diferencia"] = f"{diferencia:.1f}"
                
                # Actualizar grupo 1snTTT
                results["grupo_1sn_ttt"] = format_temperature_group(temp_seco)
                
        except ValueError:
            pass
    
    # =========================================================================
    # CÁLCULOS DE PRESIÓN
    # =========================================================================
    pres_est_str = data.get("pres_est", "")
    p3_str = data.get("p3", "")
    p24_str = data.get("p24", "")
    correc_alt_str = results["correc_alt"] or data.get("correc_alt", "")
    
    if pres_est_str:
        try:
            pres_est = float(pres_est_str)
            
            # Let. = Pres. Est. (auto-llenado)
            results["p3_let"] = f"{pres_est:.1f}"
            results["p24_let"] = f"{pres_est:.1f}"
            
            # P3 Dif. = Pres. Est. - P3
            if p3_str:
                try:
                    p3 = float(p3_str)
                    p3_dif = pres_est - p3
                    results["p3_dif"] = f"{p3_dif:.1f}"
                    
                    # Generar grupo 5aPPP (tendencia de presión)
                    results["grupo_5appp"] = format_pressure_tendency_group(p3_dif)
                except ValueError:
                    pass
            
            # P24 Dif. = Pres. Est. - P24
            if p24_str:
                try:
                    p24 = float(p24_str)
                    p24_dif = pres_est - p24
                    results["p24_dif"] = f"{p24_dif:.1f}"
                    
                    # Generar grupo 58/59 P24P24P24 (cambio 24h)
                    results["grupo_58_59_p24"] = format_pressure_24h_group(p24_dif)
                except ValueError:
                    pass
            
            # Pres. NMM = Pres. Est. + Correc. Alt.
            if correc_alt_str:
                try:
                    correc_alt = float(correc_alt_str)
                    pres_nmm = pres_est + correc_alt
                    results["pres_nmm"] = f"{pres_nmm:.1f}"
                    # Generar grupo 4PPPP
                    results["grupo_4pppp"] = format_pressure_group(pres_nmm)
                except ValueError:
                    pass
                    
        except ValueError:
            pass
    
    # =========================================================================
    # INDICADORES Ir/Ix
    # =========================================================================
    ir = data.get("ir", "")
    ix = data.get("ix", "")
    
    results["include_precipitation"] = should_include_precipitation(ir)
    results["include_weather"] = should_include_weather(ix)
    
    return results


def get_all_stations() -> Dict[str, Dict]:
    """Retorna todas las estaciones disponibles."""
    return STATIONS

File: frontend/src-tauri/src/calculations.rs
use crate::db::DbPool;
use serde::{Deserialize, Serialize};
use crate::repositories::stations_repo::{self, StationInfo};

pub fn get_station_info(pool: &DbPool, station_id: &str) -> Option<StationInfo> {
    stations_repo::get_station_by_id(pool, station_id).unwrap_or(None)
}

// =============================================================================
// CÁLCULOS METEOROLÓGICOS
// =============================================================================

fn calcular_h26(temp_humedo: f64) -> Option<f64> {
    if (temp_humedo + 243.5).abs() < f64::EPSILON {
        return None;
    }
    Some(6.112 * ((17.67 * temp_humedo) / (temp_humedo + 243.5)).exp())
}

fn calcular_h27(temp_humedo: f64) -> f64 {
    0.00066 * 1000.0 * (1.0 + 0.00115 * temp_humedo)
}

fn calcular_h28_asumido(temp_seco: f64) -> Option<f64> {
    if (temp_seco + 243.5).abs() < f64::EPSILON {
        return None;
    }
    Some(6.112 * ((17.67 * temp_seco) / (temp_seco + 243.5)).exp())
}

fn calcular_tension_vapor(temp_seco: f64, temp_humedo: f64) -> Option<f64> {
    let h26 = calcular_h26(temp_humedo)?;
    let h27 = calcular_h27(temp_humedo);
    Some(h26 - h27 * (temp_seco - temp_humedo))
}

fn calcular_humedad_relativa(tension_vapor: f64, temp_seco: f64) -> Option<f64> {
    let h28_asumido = calcular_h28_asumido(temp_seco)?;
    if h28_asumido == 0.0 {
        return None;
    }
    Some((tension_vapor / h28_asumido) * 100.0)
}

fn calcular_punto_rocio(temp_seco: f64, humedad_relativa_porc: f64) -> Option<f64> {
    let factor_humedad = 1.0 - (0.01 * humedad_relativa_porc);
    let termino1 = (14.55 + 0.114 * temp_seco) * factor_humedad;
    let termino2 = ((2.5 + 0.007 * temp_seco) * factor_humedad).powf(3.0);
    let termino3 = (15.9 + 0.117 * temp_seco) * factor_humedad.powf(14.0);
    let result = temp_seco - termino1 - termino2 - termino3;
    if result.is_nan() || result.is_infinite() {
        None
    } else {
        Some(result)
    }
}

// =============================================================================
// CODIFICACIÓN SYNOP (WMO FM-12)
// =============================================================================

fn format_temperature_group(temp: Option<f64>) -> String {
    if let Some(t) = temp {
        let sign = if t >= 0.0 { "0" } else { "1" };
        let ttt = format!("{:03}", (t.abs() * 10.0).round() as i32);
        format!("1{}{}", sign, ttt)
    } else {
        "".to_string()
    }
}

fn format_dew_point_group(dew_point: Option<f64>) -> String {
    if let Some(dp) = dew_point {
        let sign = if dp >= 0.0 { "0" } else { "1" };
        let ttt = format!("{:03}", (dp.abs() * 10.0).round() as i32);
        format!("2{}{}", sign, ttt)
    } else {
        "".to_string()
    }
}

fn format_pressure_group(pressure: Option<f64>) -> String {
    if let Some(p) = pressure {
        let pppp = format!("{:04}", ((p * 10.0).round() as i32) % 10000);
        format!("4{}", pppp)
    } else {
        "".to_string()
    }
}

fn calcular_tendencia_a(dif: f64) -> String {
    let abs_dif = dif.abs();

    if dif == 0.0 {
        return "4".to_string();
    }

    if dif > 0.0 {
        if (0.1..=0.5).contains(&abs_dif) {
            return "0".to_string();
        }
        if (0.6..=1.4).contains(&abs_dif) {
            return "1".to_string();
        }
        if (1.5..=1.9).contains(&abs_dif) {
            return "2".to_string();
        }
        if abs_dif >= 2.0 {
            return "3".to_string();
        }
    }

    if dif < 0.0 {
        if (0.1..=0.5).contains(&abs_dif) {
            return "5".to_string();
        }
        if (0.6..=1.4).contains(&abs_dif) {
            return "6".to_string();
        }
        if (1.5..=1.9).contains(&abs_dif) {
            return "7".to_string();
        }
        if abs_dif >= 2.0 {
            return "8".to_string();
        }
    }

    "4".to_string()
}

fn format_pressure_tendency_group(diferencia_p3: Option<f64>) -> String {
    if let Some(dif) = diferencia_p3 {
        let dif_rounded = (dif * 10.0).round() / 10.0;
        let a = calcular_tendencia_a(dif_rounded);
        let ppp = format!("{:03}", (dif_rounded.abs() * 10.0).round() as i32);
        format!("5{}{}", a, ppp)
    } else {
        "".to_string()
    }
}

fn format_pressure_24h_group(diferencia_p24: Option<f64>) -> String {
    if let Some(dif) = diferencia_p24 {
        let dif_rounded = (dif * 10.0).round() / 10.0;
        let prefix = if dif_rounded >= 0.0 { "58" } else { "59" };
        let ppp = format!("{:03}", (dif_rounded.abs() * 10.0).round() as i32);
        format!("{}{}", prefix, ppp)
    } else {
        "".to_string()
    }
}

fn format_humidity_group(humedad_relativa: Option<f64>) -> String {
    if let Some(hr) = humedad_relativa {
        let val = hr.clamp(0.0, 100.0).round() as i32;
        format!("29{:03}", val)
    } else {
        "".to_string()
    }
}

fn should_include_precipitation(ir: &str) -> bool {
    matches!(ir, "0" | "1" | "2")
}

fn should_include_weather(ix: &str) -> bool {
    matches!(ix, "1" | "4")
}

// =============================================================================
// I/O STRUCTURES
// =============================================================================

#[derive(Deserialize, Debug)]
pub struct CalculationRequest {
    pub station_id: Option<String>,
    pub correc_alt: Option<String>,
    pub ts: Option<String>,
    pub th: Option<String>,
    pub pres_est: Option<String>,
    pub p3: Option<String>,
    pub p24: Option<String>,
    pub ir: Option<String>,
    pub ix: Option<String>,
}

#[derive(Serialize)]
pub struct CalculationResponse {
    pub tension_vapor: String,
    pub humedad_relativa: String,
    pub punto_rocio: String,
    pub diferencia: String,

    pub grupo_1sn_ttt: String,
    pub grupo_2sn_td: String,
    pub grupo_4pppp: String,
    pub grupo_5appp: String,
    pub grupo_58_59_p24: String,
    pub grupo_29uuu: String,

    pub p3_let: String,
    pub p24_let: String,
    pub p3_dif: String,
    pub p24_dif: String,
    pub pres_nmm: String,
    pub correc_alt: String,

    pub include_precipitation: bool,
    pub include_weather: bool,

    pub station_info: Option<StationInfo>,
    pub error_message: String,
}

// =============================================================================
// ALGORITMO PRINCIPAL
// =============================================================================

pub fn realizar_calculos(pool: &DbPool, data: CalculationRequest) -> CalculationResponse {
    let mut res = CalculationResponse {
        tension_vapor: "".to_string(),
        humedad_relativa: "".to_string(),
        punto_rocio: "".to_string(),
        diferencia: "".to_string(),
        grupo_1sn_ttt: "".to_string(),
        grupo_2sn_td: "".to_string(),
        grupo_4pppp: "".to_string(),
        grupo_5appp: "".to_string(),
        grupo_58_59_p24: "".to_string(),
        grupo_29uuu: "".to_string(),
        p3_let: "".to_string(),
        p24_let: "".to_string(),
        p3_dif: "".to_string(),
        p24_dif: "".to_string(),
        pres_nmm: "".to_string(),
        correc_alt: "".to_string(),
        include_precipitation: true,
        include_weather: true,
        station_info: None,
        error_message: "".to_string(),
    };

    if let Some(st_id) = &data.station_id {
        if let Some(info) = get_station_info(pool, st_id) {
            res.correc_alt = info.ch.to_string();
            res.station_info = Some(info);
        }
    }
    if res.correc_alt.is_empty() {
        if let Some(ca) = &data.correc_alt {
            res.correc_alt = ca.clone();
        }
    }

    let ts_val = data.ts.as_ref().and_then(|s| s.parse::<f64>().ok());
    let th_val = data.th.as_ref().and_then(|s| s.parse::<f64>().ok());

    if let Some(ts) = ts_val {
        res.grupo_1sn_ttt = format_temperature_group(Some(ts));
    }

    if let (Some(ts), Some(th)) = (ts_val, th_val) {
        if th > ts {
            res.error_message = "Error: Th no puede ser mayor que Ts.".to_string();
        } else {
            if let Some(tv) = calcular_tension_vapor(ts, th) {
                res.tension_vapor = format!("{:.1}", tv);
                if let Some(hr) = calcular_humedad_relativa(tv, ts) {
                    res.humedad_relativa = format!("{:.0}", hr.clamp(0.0, 100.0));
                    if let Some(pr) = calcular_punto_rocio(ts, hr) {
                        res.punto_rocio = format!("{:.1}", pr);
                        res.grupo_2sn_td = format_dew_point_group(Some(pr));
                    }
                    res.grupo_29uuu = format_humidity_group(Some(hr));
                }
            }
            res.diferencia = format!("{:.1}", ts - th);
            res.grupo_1sn_ttt = format_temperature_group(Some(ts));
        }
    }

    let pres_est_val = data.pres_est.as_ref().and_then(|s| s.parse::<f64>().ok());
    let p3_val = data.p3.as_ref().and_then(|s| s.parse::<f64>().ok());
    let p24_val = data.p24.as_ref().and_then(|s| s.parse::<f64>().ok());
    let correc_alt_val = res.correc_alt.parse::<f64>().ok();

    if let Some(pe) = pres_est_val {
        res.p3_let = format!("{:.1}", pe);
        res.p24_let = format!("{:.1}", pe);

        if let Some(p3) = p3_val {
            let dif = pe - p3;
            res.p3_dif = format!("{:.1}", dif);
            res.grupo_5appp = format_pressure_tendency_group(Some(dif));
        }

        if let Some(p24) = p24_val {
            let dif = pe - p24;
            res.p24_dif = format!("{:.1}", dif);
            res.grupo_58_59_p24 = format_pressure_24h_group(Some(dif));
        }

        if let Some(ca) = correc_alt_val {
            let p_nmm = pe + ca;
            res.pres_nmm = format!("{:.1}", p_nmm);
            res.grupo_4pppp = format_pressure_group(Some(p_nmm));
        }
    }

    res.include_precipitation = data
        .ir
        .as_ref()
        .map(|s| should_include_precipitation(s))
        .unwrap_or(true);
    res.include_weather = data
        .ix
        .as_ref()
        .map(|s| should_include_weather(s))
        .unwrap_or(true);

    res
}

use std::collections::HashMap;

#[tauri::command]
pub fn get_stations(pool: tauri::State<'_, DbPool>) -> Result<HashMap<String, StationInfo>, String> {
    let stations = stations_repo::get_all_stations(&pool)?;
    let mut map = HashMap::new();
    for st in stations {
        map.insert(st.id.clone(), st);
    }
    Ok(map)
}

#[tauri::command]
pub fn create_station(pool: tauri::State<'_, DbPool>, station: StationInfo) -> Result<(), String> {
    stations_repo::insert_station(&pool, &station)
}

#[tauri::command]
pub fn update_station(pool: tauri::State<'_, DbPool>, station: StationInfo) -> Result<(), String> {
    stations_repo::update_station(&pool, &station)
}

#[tauri::command]
pub fn delete_station(pool: tauri::State<'_, DbPool>, id: String) -> Result<(), String> {
    stations_repo::delete_station(&pool, &id)
}

#[tauri::command]
pub fn calculate_observations(pool: tauri::State<'_, DbPool>, data: CalculationRequest) -> CalculationResponse {
    realizar_calculos(&pool, data)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_tension_vapor() {
        let ts = 25.0;
        let th = 20.0;
        let tv = calcular_tension_vapor(ts, th).unwrap();
        assert!((tv - 19.3).abs() < 1.0, "Tensión de vapor inválida: {}", tv);
    }

    #[test]
    fn test_humedad_relativa() {
        let ts = 25.0;
        let tv = 19.3;
        let hr = calcular_humedad_relativa(tv, ts).unwrap();
        assert!(hr > 50.0 && hr < 70.0, "Humedad relativa inválida: {}", hr);
    }

    #[test]
    fn test_punto_rocio() {
        let ts = 25.0;
        let hr = 60.0;
        let pr = calcular_punto_rocio(ts, hr).unwrap();
        assert!((pr - 16.7).abs() < 2.0, "Punto de rocío inválido: {}", pr);
    }

    #[test]
    fn test_tendencia_a() {
        assert_eq!(calcular_tendencia_a(0.0), "4");
        assert_eq!(calcular_tendencia_a(0.3), "0");
        assert_eq!(calcular_tendencia_a(1.0), "1");
        assert_eq!(calcular_tendencia_a(1.8), "2");
        assert_eq!(calcular_tendencia_a(2.5), "3");
        assert_eq!(calcular_tendencia_a(-0.3), "5");
        assert_eq!(calcular_tendencia_a(-1.0), "6");
        assert_eq!(calcular_tendencia_a(-1.8), "7");
        assert_eq!(calcular_tendencia_a(-2.5), "8");
    }

    #[test]
    fn test_format_temperature_group() {
        assert_eq!(format_temperature_group(Some(25.4)), "10254");
        assert_eq!(format_temperature_group(Some(-5.2)), "11052");
        assert_eq!(format_temperature_group(None), "");
    }

    #[test]
    fn test_format_dew_point_group() {
        assert_eq!(format_dew_point_group(Some(16.7)), "20167");
        assert_eq!(format_dew_point_group(Some(-3.4)), "21034");
        assert_eq!(format_dew_point_group(None), "");
    }

    #[test]
    fn test_format_pressure_group() {
        assert_eq!(format_pressure_group(Some(1015.3)), "40153");
        assert_eq!(format_pressure_group(Some(998.4)), "49984");
        assert_eq!(format_pressure_group(None), "");
    }
}
